src/transformers/models/xcodec2/modular_xcodec2.py

```diff
@@ -0,0 +1,1686 @@
+            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+        elif isinstance(module, (nn.Conv1d, nn.ConvTranspose1d)):
+            nn.init.kaiming_normal_(module.weight)
+            if module.bias is not None:
+                k = math.sqrt(module.groups / (module.in_channels * module.kernel_size[0]))
+                nn.init.uniform_(module.bias, a=-k, b=k)
+
+
+XCODEC2_START_DOCSTRING = r"""
+    This model inherits from [`PreTrainedModel`]. Check the superclass documentation for the generic methods the
+    library implements for all its model (such as downloading or saving, resizing the input embeddings, pruning heads
+    etc.)
+    This model is also a PyTorch [torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) subclass.
+    Use it as a regular PyTorch Module and refer to the PyTorch documentation for all matter related to general usage
+    and behavior.
+    Parameters:
+        config ([`Xcodec2Config`]):
+            Model configuration class with all the parameters of the model. Initializing with a config file does not
+            load the weights associated with the model, only the configuration. Check out the
+            [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+XCODEC2_INPUTS_DOCSTRING = r"""
+    args:
+        input_values (`torch.FloatTensor` of shape `(batch_size, channels, num_samples)`):
+            The raw float values of the input audio waveform.
+        audio_codes (`torch.LongTensor`  of shape `(batch_size, num_quantizers, codes_length)`:
+            Discrete code indices computed using `model.encode`.
+        return_dict (`bool`, *optional*):
+            whether to return a `XcodecOutput` or a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    "The Xcodec2 neural audio codec model.",
+    XCODEC2_INPUTS_DOCSTRING,
+)
+class XCodec2Model(XCodec2PreTrainedModel):
+    config_class = XCodec2Config
+
+    def __init__(self, config: XCodec2Config):
+        super().__init__(config)
+
+        semantic_model_config = AutoConfig.from_pretrained("facebook/w2v-bert-2.0", output_hidden_states=True)
+        self.semantic_model = AutoModel.from_config(semantic_model_config)
+        self.semantic_model.eval()
+
+        self.SemanticEncoder_module = XCodec2SemanticEncoder(
+            config.semantic_hidden_size, config.semantic_hidden_size, config.semantic_hidden_size
+        )
+
+        self.CodecEnc = XCodec2CodecEncoder_Transformer()
+
+        self.generator = XCodec2CodecDecoderVocos(config=config)
+
+        self.fc_prior = nn.Linear(2048, 2048)
+        self.fc_post_a = nn.Linear(2048, 1024)
+        feature_extractor = AutoFeatureExtractor.from_pretrained("facebook/w2v-bert-2.0")
```

Comment:
@eustlb should loading other checkpoints with `from_pretrained` be wrapped in something like a processor? or fine to use inside modeling code?